dags/iop/parser.py

```diff
@@ -106,6 +117,51 @@ def _get_extracted_arxiv_eprint_value(self, article):
         except AttributeError:
             self.logger.error("No arXiv eprints found", dois=self.dois)
 
+    def _get_date_element(self, article):
+        try:
+            date = (
+                article.find("front/article-meta/history/date[@date-type='published']")
+                or article.find("front/article-meta/history/date[@date-type='epub']")
+                or article.find("front/article-meta/history/pub-date[@pub-type='ppub']")
+                or article.find("front/article-meta/pub-date")
+            )
+            return date
+        except AttributeError:
+            return
+
+    def _get_date_published(self, article):
+        date_element = self._get_date_element(article)
+        return self._get_date(date_element)
+
+    def _get_date(self, date):
+        day = None
+        month = None
+        try:
+            year = int(date.find("year").text)
+            self.year = year
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find year of date_published in XML")
+        try:
+            month = int(date.find("month").text)
+        # was discussed with Anne: if the day is not present, we return year and month
+        # if the month value is not present - we return just year
+        # if the year is not present - the parsing should crash, because according schema
+        # the publication_year is required, which is take from date_published
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find month of date_published in XML")
+        try:
+            day = int(date.find("day").text)
+        except (AttributeError, TypeError):
+            self.logger.error("Cannot find day of date_published in XML")
+        try:
+            date = PartialDate(year=self.year, month=month, day=day)
+            return PartialDate.dumps(date)
+        except (AttributeError, ValueError, TypeError):
+            return str(self.year)
+
+    def _get_journal_year(self, _):
+        return self.year
```

Comment:
do the extractors run in a specific order? will `self.year` be there if `_get_journal_year` runs before `_get_date_published`

Reply:
yes, if it is in XML. Extractor runs in the order as it in the extractors array:
https://github.com/cern-sis/workflows/blob/date_published/dags/common/parsing/parser.py#L40-L41